### DISCUSSION TITLE
Does memory need to be mapped aligned to page boundaries?

### DISCUSSION THREAD

**Discussion by Joshua-Schroijen** [unicorn-engine/unicorn#1948] [Q&A] [answered]
Hello!  
I have a very simple question. A lot of Unicorn code I have been studying seems to make sure the start and end of memory regions to be mapped in the emulator using `mem_map()` are aligned to page boundaries. Is this necessary or a good idea (and if so, why)? Could failing to do this cause strange behavior?

Is there anything `mem_map()` explicitly or implicitly requires about its memory addresses to be aware about at all? Or are arbitrary byte-level addresses OK?

Thanks in advance for the input!  
Joshua

**Comment by wtdcode** [comment]
This is usually decided by MMU. For example, many MMU design & implementation will assume 4k pages and use the low free bits.

**Reply by Joshua-Schroijen** [reply]
Maybe I understand it poorly but that is the actual core of my question. Obivously an MMU is going to intimately care about paging, but does mem_map do anything with paging? Doesn't that just make memory available to the emulator? Does it do anything more, involving pages?

**Reply by wtdcode** [reply, marked as answer]
Hmmmm, that's quite subtle. Though not thoroughly confirmed, I think the alignment requirement is a QEMU internal implementation trick, which we just inherited. I don't see any straight reason making it possible to support any memory regions alignments but meanwhile there is also no obvious motivation.
